File: indirect_gwas/igwas.py
import argparse
import datetime
import os
import pathlib

import numpy as np
import pandas as pd
import polars as pl
import scipy.stats


class IndirectGWAS:
    def __init__(
        self,
        gwas_summary_statistics: list[os.PathLike],
        projection_coefficients: np.ndarray | pd.DataFrame,
        feature_partial_covariance: np.ndarray | pd.DataFrame,
        n_exogenous: int,
        output: os.PathLike,
        chunksize: int = 100_000,
        variant_id_column: str = "ID",
        beta_column: str = "BETA",
        se_column: str = "SE",
        n_column: str = "OBS_CT",
        computation_dtype: str = "float32",
        separator: str = "\t",
        quiet: bool = False,
    ):
        # Indirect GWAS data
        self.gwas_summary_statistics = gwas_summary_statistics
        self.projection_coefficients = projection_coefficients
        self.covariance = feature_partial_covariance
        self.n_exogenous = n_exogenous

        # Options
        self.output = pathlib.Path(output)
        self.output = (
            self.output.with_suffix(".tsv") if self.output.suffix == "" else self.output
        )
        self.chunksize = chunksize
        self.computation_dtype = computation_dtype
        self.separator = separator
        self.quiet = quiet

        # Column names
        self.variant_id_column = variant_id_column
        self.beta_column = beta_column
        self.se_column = se_column
        self.n_column = n_column
        self.zfill = (np.floor(np.log10(len(self.gwas_summary_statistics))) + 1).astype(
            int
        )

        # Check inputs and gather statistics
        self.projection_ids, self.variant_ids = self.check_inputs()

        # Derived quantities
        self.n_features = len(gwas_summary_statistics)
        self.n_projections = len(self.projection_ids)
        self.n_variants = len(self.variant_ids)
        self.feature_partial_variance = np.diag(feature_partial_covariance)
        self.var_p_p = np.diag(
            projection_coefficients.T
            @ feature_partial_covariance
            @ projection_coefficients
        ).reshape(-1, 1)
        self.memory_map_temporary_path = None
        self.mmap = None

        if not self.quiet:
            print(f"Found {self.n_features} features")
            print(f"Found {self.n_projections} projections")
            print(f"Found {self.n_variants} variants")

    def check_inputs(self):
        # Check that all gwas_sumstats files exist
        for path in self.gwas_summary_statistics:
            if not os.path.exists(path):
                raise FileNotFoundError(f"File not found: {path}")

        # Check that all gwas_sumstat files have exactly the same variant IDs in the
        # same order
        variant_ids = None
        for path in self.gwas_summary_statistics:
            file_variant_ids = pl.read_csv(
                path, separator=self.separator, columns=[self.variant_id_column]
            )[self.variant_id_column].to_numpy()
            if variant_ids is None:
                variant_ids = file_variant_ids
            else:
                if not np.array_equal(variant_ids, file_variant_ids):
                    raise ValueError(
                        f"Variant IDs in {path} do not match those in the other files"
                    )

        # Check that the projection coefficients have the same number of rows as the
        # number of files in gwas_sumstats
        if isinstance(self.projection_coefficients, pd.DataFrame | np.ndarray):
            if self.projection_coefficients.shape[0] != len(
                self.gwas_summary_statistics
            ):
                raise ValueError(
                    "Number of rows in projection coefficients does not match the "
                    "number of files in gwas_sumstats"
                )
        else:
            raise TypeError("projection_coefficients must be an array or DataFrame")

        # Check that the feature partial covariance has the same number of rows as the
        # number of files in gwas_sumstats
        if isinstance(self.covariance, pd.DataFrame | np.ndarray):
            if self.covariance.shape[0] != len(self.gwas_summary_statistics):
                raise ValueError(
                    "Number of rows in feature partial covariance does not match the "
                    "number of files in gwas_sumstats"
                )
        else:
            raise TypeError("feature_partial_covariance must be an array or DataFrame")

        # Get the projection names
        if isinstance(self.projection_coefficients, pd.DataFrame):
            projection_ids = self.projection_coefficients.columns
        else:
            projection_ids = [
                f"projection_{str(i).zfill(self.zfill)}"
                for i in range(self.projection_coefficients.shape[1])
            ]

        return projection_ids, variant_ids

    def create_mmap_file(self):
        """
        Shape is (6, n_projections, n_variants)
        0: degrees-of-freedom
        1: beta
        2: sum of genotype partial variances
        3: standard error
        4: t-statistic
        5: p-value
        """
        self.memory_map_temporary_path = self.output.parent.joinpath(
            f"{self.output.stem}_temp.mmap"
        )

        self.mmap = np.memmap(
            self.memory_map_temporary_path,
            dtype=self.computation_dtype,
            mode="w+",
            shape=(6, self.n_projections, self.n_variants),
        )
        self.mmap[0] = np.inf
        self.mmap[1:] = 0.0

    def process_file_chunked(
        self,
        feature_gwas_path: os.PathLike,
        feature_projection_coefficients: np.ndarray,
        feature_partial_variance: float,
    ):
        reader = pl.read_csv_batched(
            feature_gwas_path,
            separator=self.separator,
            has_header=True,
            batch_size=self.chunksize,
            columns=["BETA", "SE", "OBS_CT"],
        )
        last_index = 0
        while True:
            chunks = reader.next_batches(1)
            if chunks is None:
                break

            chunk = chunks[0]

            beta = chunk["BETA"].to_numpy().astype(self.computation_dtype)
            se = chunk["SE"].to_numpy().astype(self.computation_dtype)
            N = chunk["OBS_CT"].to_numpy().astype(self.computation_dtype)

            current_chunk = self.mmap[:, :, last_index : last_index + len(chunk)]

            # Update the degrees-of-freedom
            dof = N - self.n_exogenous - 1
            current_chunk[0] = np.minimum(current_chunk[0], dof)

            # Update beta
            current_chunk[1] += np.outer(feature_projection_coefficients, beta)

            # Update the sum of genotype partial variances
            current_chunk[2] += feature_partial_variance / (dof * se**2 + beta**2)
            last_index += len(chunk)

        self.mmap.flush()

    def make_chunk_df(self, chunk: np.memmap, variant_ids: list[str]):
        output_df = pd.DataFrame(
            {
                "projection": np.tile(self.projection_ids, chunk.shape[2]),
                "variant": np.repeat(variant_ids, self.n_projections),
                "beta": chunk[1].ravel("F"),
                "se": chunk[3].ravel("F"),
                "t": chunk[4].ravel("F"),
                "nlog10p": chunk[5].ravel("F"),
                "dof": chunk[0].ravel("F"),
            }
        )
        return output_df

    def write_output_chunked(self):
        chunk_indices = list(range(0, self.n_variants, self.chunksize))
        for chunk_idx in chunk_indices:
            chunk = self.mmap[:, :, chunk_idx : chunk_idx + self.chunksize]
            variant_ids = self.variant_ids[chunk_idx : chunk_idx + self.chunksize]

            output_df = self.make_chunk_df(chunk, variant_ids)
            for name, group_df in output_df.groupby("projection"):
                output_path = self.output.with_name(
                    f"{self.output.stem}_{name}{self.output.suffix}"
                )
                group_df.drop(columns=["projection"]).to_csv(
                    output_path,
                    mode=("w" if chunk_idx == 0 else "a"),
                    header=(chunk_idx == 0),
                    sep="\t",
                    index=False,
                    chunksize=100_000,
                )

    def run(self):
        # Create the memory map temporary file
        if not self.quiet:
            print("Opening a temporary file to store computations")

        if self.mmap is None:
            self.create_mmap_file()

        # Process each GWAS summary statistic file, adding the results to the memory map
        for i, path in enumerate(self.gwas_summary_statistics):
            if not self.quiet:
                name = str(i + 1).zfill(self.zfill)
                print(
                    f"Processing file {name}/{len(self.gwas_summary_statistics)}: {path}"
                )

            self.process_file_chunked(
                path,
                np.array(self.projection_coefficients)[i],
                np.array(self.feature_partial_variance)[i],
            )

        # Compute summary statistics
        if not self.quiet:
            print("Computing final summary statistics")
        # Genotypic partial variance is taken as the mean across features
        self.mmap[2] = self.mmap[2] / self.n_features
        # Standard error
        self.mmap[3] = np.sqrt(
            (self.var_p_p / self.mmap[2] - self.mmap[1] ** 2) / self.mmap[0]
        )
        # t-statistic
        self.mmap[4] = self.mmap[1] / self.mmap[3]
        # p-value (-log10(2 * sf) is -1 * (log(sf) + log(2)) / log(10)
        self.mmap[5] = (
            -1
            * (scipy.stats.t.logsf(np.abs(self.mmap[4]), self.mmap[0]) + np.log(2))
            / np.log(10)
        )

        self.mmap.flush()

        # Write the output
        if not self.quiet:
            print("Completed computations, generating output file(s)")

        self.write_output_chunked()

        # Clean up the memory map temporary file
        if not self.quiet:
            print("Finished writing output, cleaning up temporary file")

        os.remove(self.memory_map_temporary_path)

File: indirect_gwas/src/igwas.cpp
#include "igwas.hpp"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <queue>
#include <mutex>
#include <condition_variable>
#include <sstream>
#include <thread>
#include <string>
#include <vector>

#include <csv.hpp>
#include <Eigen/Dense>

#include "io.hpp"
#include "utils.hpp"

IndirectGWAS::IndirectGWAS(
    ColumnSpec column_names,
    LabeledMatrix projection_coefficients,
    LabeledMatrix feature_partial_covariance,
    const unsigned int n_covariates,
    const unsigned int chunksize,
    bool single_file_output) : column_names(column_names),
                               n_covariates(n_covariates),
                               chunksize(chunksize),
                               projection_coefficients(projection_coefficients),
                               single_file_output(single_file_output)
{
    auto &projection_names = projection_coefficients.column_names;
    auto &feature_names = feature_partial_covariance.row_names;

    // Check that the features are the same in both files
    ensure_names_consistent(projection_coefficients.row_names, feature_names);

    // Extract the diagonal of the feature partial covariance matrix
    auto &C = feature_partial_covariance.data;
    feature_partial_variance = {C.diagonal(), feature_names};

    // Compute the partial variance of each projection
    auto &P = projection_coefficients.data;
    projection_partial_variance = {(P.transpose() * C * P).diagonal(), projection_names};

    // Compute the number of features and projections
    n_features = feature_names.size();
    n_projections = projection_names.size();

    // Initialize the chunk matrices
    degrees_of_freedom = {Eigen::VectorXf::Zero(0), variant_ids};
    beta = {Eigen::MatrixXf::Zero(0, n_projections), variant_ids, projection_names};
    gpv_sum = {Eigen::VectorXf::Zero(0), variant_ids};
}

void IndirectGWAS::ensure_names_consistent(std::vector<std::string> names_1,
                                           std::vector<std::string> names_2)
{
    if (names_1 != names_2)
    {
        std::cerr << "Error: feature names do not match between files" << std::endl;

        std::cerr << "First group: ";
        for (auto name : names_1)
            std::cerr << name << " ";
        std::cerr << std::endl;

        std::cerr << "Second group: ";
        for (auto name : names_2)
            std::cerr << name << " ";
        std::cerr << std::endl;

        throw std::runtime_error("Feature names do not match");
    }
}

void IndirectGWAS::process_file_chunk(unsigned int k, FeatureGwasResults &results)
{
    if (k == 0)
    {
        variant_ids = results.variant_ids;
        degrees_of_freedom.data = results.degrees_of_freedom;
    }
    else
    {
        // Check that the variant IDs match
        ensure_names_consistent(variant_ids, results.variant_ids);
        // Update the degrees of freedom
        degrees_of_freedom.data = degrees_of_freedom.data.cwiseMin(results.degrees_of_freedom);
    }

    // Update the beta matrix
    for (int i = 0; i < variant_ids.size(); i++)
    {
        for (int j = 0; j < projection_coefficients.row_names.size(); j++)
        {
            beta.data(i, j) += results.beta(i) * projection_coefficients.data(k, j);
        }
    }

    // Update the sum of genotype partial variances
    Eigen::VectorXf &dof = results.degrees_of_freedom;
    Eigen::VectorXf &se = results.std_error;
    Eigen::VectorXf &b = results.beta;

    Eigen::VectorXf denom = dof.array() * se.array().square() + b.array().square();
    gpv_sum.data += (denom.cwiseInverse() * feature_partial_variance.data[k]);
}

// Split the standard error computation into a separate function for readability
void IndirectGWAS::compute_standard_error(ResultsChunk &results)
{
    // Create a reference to gpv_sum that is called gpv_mean (to avoid ambiguity)
    Eigen::VectorXf &gpv = gpv_sum.data;
    gpv /= n_features;

    // Reference to the standard error matrix
    Eigen::MatrixXf &se = results.std_error;

    se = -results.beta.array().square();
    for (int i = 0; i < se.rows(); i++)
    {
        for (int j = 0; j < se.cols(); j++)
        {
            se(i, j) += projection_partial_variance.data(j) / gpv(i);
            se(i, j) /= degrees_of_freedom.data(i);

            if (std::isnan(se(i, j)) || se(i, j) < 0)
            {
                std::cerr << "Standard error for variant "
                          << results.variant_ids[i] << " is " << se(i, j)
                          << ". gpv = " << gpv(i)
                          << ", projection_partial_variance = "
                          << projection_partial_variance.data(j)
                          << ", degrees_of_freedom = " << degrees_of_freedom.data(i)
                          << ", beta = " << results.beta(i, j) << std::endl;
            }
        }
    }
    se = se.cwiseSqrt();
}

void IndirectGWAS::compute_p_value(ResultsChunk &results)
{
    Eigen::MatrixXf &t = results.t_statistic;
    Eigen::VectorXf &dof = degrees_of_freedom.data;
    Eigen::MatrixXf &p = results.neg_log10_p_value;

    p.resizeLike(t);
    for (int i = 0; i < t.rows(); i++)
    {
        for (int j = 0; j < t.cols(); j++)
        {
            if (dof(i) <= 0 || std::isnan(dof(i)))
            {
                throw std::runtime_error("Degrees of freedom is an error for variant " +
                                         results.variant_ids[i] + " with value " +
                                         std::to_string(dof(i)));
            }
            if (std::isnan(t(i, j)))
            {
                std::cerr << "Erroneous T-statistic for variant "
                          << results.variant_ids[i] << " with value " << t(i, j)
                          << ". Corresponding beta is " << results.beta(i, j)
                          << ", standard error is " << results.std_error(i, j)
                          << std::endl;

                p(i, j) = std::nan("");
                continue;
            }
            p(i, j) = compute_log_p_value(t(i, j), dof(i));
        }
    }
}

ResultsChunk IndirectGWAS::compute_results_chunk()
{
    ResultsChunk results;

    results.variant_ids = std::move(variant_ids);
    results.beta = std::move(beta.data);

    compute_standard_error(results);

    // Compute the t-statistic
    results.t_statistic = results.beta.array() / results.std_error.array();

    // Compute the p-value by calling compute_log_p_value on each t-statistic
    compute_p_value(results);

    // Add the sample size to the results
    results.sample_size = degrees_of_freedom.data.array() + n_covariates + 2;

    return results;
};

void IndirectGWAS::save_results_chunk(ResultsChunk &results, std::string output_stem, bool write_header)
{
    for (int i = 0; i < results.beta.cols(); i++)
    {
        std::string projection_name = projection_coefficients.column_names[i];

        // Open the output file
        std::string filename = output_stem + "_" + projection_name + ".csv";
        std::ofstream file;

        if (write_header)
        {
            file.open(filename);

            // Write the header
            // IDEA: Could use the same column names as the input files
            file << "variant_id,beta,std_error,t_statistic,neg_log10_p_value,sample_size"
                 << std::endl;
        }
        else
        {
            file.open(filename, std::ios_base::app);
        }

        // IDEA: Could set precision as a parameter
        file << std::setprecision(6);

        // Write the results
        for (int j = 0; j < results.variant_ids.size(); j++)
        {
            // IDEA: Could use other separators
            file << results.variant_ids[j] << ","
                 << results.beta(j, i) << ","
                 << results.std_error(j, i) << ","
                 << results.t_statistic(j, i) << ","
                 << results.neg_log10_p_value(j, i) << ","
                 << results.sample_size(j) << std::endl;
        }
    }
}

void IndirectGWAS::save_results_single_file(ResultsChunk &results, std::string output_stem, bool write_header)
{
    int vidSize = results.variant_ids.size();
    int pidSize = results.beta.cols();

    std::ios_base::sync_with_stdio(false);

    // Preallocate a string to hold the results. Expect ballpark 70 characters per line
    // and number of lines equal to the number of variants times the number of projections
    std::string str;
    str.reserve(70 * vidSize * pidSize);

    std::ostringstream oss(std::move(str));
    oss.precision(6);

    for (int vid = 0; vid < vidSize; vid++)
    {
        for (int pid = 0; pid < pidSize; pid++)
        {
            oss << projection_coefficients.column_names[pid] << ","
                << results.variant_ids[vid] << ","
                << results.beta(vid, pid) << ","
                << results.std_error(vid, pid) << ","
                << results.t_statistic(vid, pid) << ","
                << results.neg_log10_p_value(vid, pid) << ","
                << results.sample_size(vid) << "\n";
        }
    }
    // Open the output file
    std::string filename = output_stem + ".csv";
    std::ofstream file;

    if (write_header)
    {
        file.open(filename);
        file << "projection_id,variant_id,beta,std_error,t_statistic,"
             << "neg_log10_p_value,sample_size" << std::endl;
    }
    else
    {
        file.open(filename, std::ios_base::app);
    }

    file << oss.str();
}

// Resets running data containers to the current chunk size and zeros where necessary
void IndirectGWAS::reset_running_data(unsigned int chunksize)
{
    // Clear the variant IDs
    variant_ids.clear();

    degrees_of_freedom.data = Eigen::VectorXf::Zero(chunksize);
    beta.data = Eigen::MatrixXf::Zero(chunksize, n_projections);
    gpv_sum.data = Eigen::VectorXf::Zero(chunksize);
}

void IndirectGWAS::load_chunk(std::vector<std::string> filenames, unsigned int chunk_start, unsigned int chunk_end)
{
    finished_reading = false;
    std::thread producer_thread(&IndirectGWAS::raw_results_producer, this, filenames,
                                chunk_start, chunk_end);
    std::thread consumer_thread(&IndirectGWAS::raw_results_consumer, this);

    producer_thread.join();
    consumer_thread.join();
}

void IndirectGWAS::raw_results_producer(
    std::vector<std::string> filenames,
    unsigned int chunk_start,
    unsigned int chunk_end)
{
    for (unsigned int i = 0; i < filenames.size(); i++)
    {
        FeatureGwasResults results = read_gwas_chunk(
            filenames[i], column_names, chunk_start, chunk_end, n_covariates);

        {
            std::lock_guard<std::mutex> lock(results_mutex);
            results_queue.push({i, results});
        }
        cv.notify_one();
    }
    finished_reading = true;
    cv.notify_all();
}

void IndirectGWAS::raw_results_consumer()
{
    while (true)
    {
        std::unique_lock<std::mutex> lock(results_mutex);
        cv.wait(lock, [this] { return !results_queue.empty() || finished_reading; });

        if (results_queue.empty() && finished_reading) break;

        IndexedGwasChunk chunk = results_queue.front();
        results_queue.pop();
        lock.unlock();

        process_file_chunk(chunk.feature_index, chunk.results);
    }
}

// Runs the indirect GWAS. Assumes that all input files are identically formatted,
// with the same column names and identical variants in the same order.
void IndirectGWAS::run(std::vector<std::string> filenames, std::string output_stem)
{
    // Create a log file for the run
    std::ofstream logfile(output_stem + ".log");

    // Add the start time to the log file
    auto start = std::chrono::high_resolution_clock::now();
    logfile << "Start time: " << get_formatted_time() << std::endl;

    // Count the number of lines to ensure appropriate chunking
    logfile << "Counting lines in " << filenames[0] << std::endl;
    unsigned int n_lines = count_lines(filenames[0]) - 1;
    logfile << "Number of lines: " << n_lines << std::endl;

    // Iterate across all chunks
    unsigned int chunk_start_line = 0;
    while (chunk_start_line < n_lines)
    {
        unsigned int chunk_end_line = std::min(chunk_start_line + chunksize - 1, n_lines - 1);

        logfile << "Processing lines " << chunk_start_line << " to " << chunk_end_line << std::endl;

        // Reset the running data to the current chunk size and zero where necessary
        reset_running_data(chunk_end_line - chunk_start_line + 1);

        // Iterate across all files, updating running summary statistics
        load_chunk(filenames, chunk_start_line, chunk_end_line);

        logfile << "Finished reading files for this chunk" << std::endl;

        // Compute the final results for this chunk
        ResultsChunk results = compute_results_chunk();

        logfile << "Finished computing final results for this chunk" << std::endl;

        // Save the results for this chunk
        if (single_file_output)
        {
            save_results_single_file(results, output_stem, chunk_start_line == 0);
        }
        else
        {
            save_results_chunk(results, output_stem, chunk_start_line == 0);
        }

        logfile << "Saved results for this chunk" << std::endl;

        // Update the chunk start line
        chunk_start_line = chunk_end_line + 1;
    }

    auto end = std::chrono::high_resolution_clock::now();

    // Add the end time to the log file
    logfile << "End time: " << get_formatted_time() << std::endl;

    // Add elapsed time
    std::chrono::duration<double> elapsed = end - start;
    logfile << "Time taken: " << elapsed.count() << " seconds" << std::endl;
}
